include/seqan3/alignment/matrix/detail/trace_matrix_full.hpp

```diff
@@ -0,0 +1,332 @@
+    //!\brief The type of the virtual score column which only stores one value.
+    using virtual_column_t = decltype(views::repeat_n(trace_t{}, 1));
+
+    class iterator;
+
+    //!\brief The full trace matrix.
+    matrix_t complete_matrix{};
+    //!\brief The column over the horizontal traces.
+    physical_column_t horizontal_column{};
+    //!\brief The virtual column over the vertical traces.
+    virtual_column_t vertical_column{};
+    //!\brief The number of columns for this matrix.
+    size_t column_count{};
+    //!\brief The number of rows for this matrix.
+    size_t row_count{};
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    trace_matrix_full() = default; //!< Defaulted.
+    trace_matrix_full(trace_matrix_full const &) = default; //!< Defaulted.
+    trace_matrix_full(trace_matrix_full &&) = default; //!< Defaulted.
+    trace_matrix_full & operator=(trace_matrix_full const &) = default; //!< Defaulted.
+    trace_matrix_full & operator=(trace_matrix_full &&) = default; //!< Defaulted.
+    ~trace_matrix_full() = default; //!< Defaulted.
+    //!\}
+
+    /*!\brief Resizes the matrix.
+     * \tparam column_index_t The column index type; must model std::integral.
+     * \tparam row_index_t The row index type; must model std::integral.
+     *
+     * \param[in] column_count The number of columns for this matrix.
+     * \param[in] row_count The number of rows for this matrix.
+     *
+     * \details
+     *
+     * Resizes the entire trace matrix storing the best trace path and the horizontal trace column.
+     * Note the trace matrix requires the number of columns and rows to be one bigger than the size of sequence1,
+     * respectively sequence2 for the initialisation of the matrix.
+     * Reallocation happens only if the new column size exceeds the current capacity of the underlying trace matrix.
+     *
+     * ### Complexity
+     *
+     * In worst case `column_count` times `row_count` memory is allocated.
+     *
+     * ### Exception
+     *
+     * Basic exception guarantee. Might throw std::bad_alloc on resizing the internal matrices.
+     */
+    template <std::integral column_index_t, std::integral row_index_t>
+    void resize(column_index_type<column_index_t> const column_count,
+                row_index_type<row_index_t> const row_count)
+    {
+        this->column_count = column_count.get();
+        this->row_count = row_count.get();
+        complete_matrix.resize(number_rows{this->row_count}, number_cols{this->column_count});
+        horizontal_column.resize(this->row_count);
+        vertical_column = views::repeat_n(trace_t{}, this->row_count);
+    }
+
+    /*!\name Iterators
+     * \{
+     */
+    //!\brief Returns the iterator pointing to the first column.
+    iterator begin()
+    {
+        return iterator{*this, 0u};
+    }
+
+    //!\brief This score matrix is not const-iterable.
+    iterator begin() const = delete;
```

Comment:
I think, it tells you that the requested function is deleted rather than telling you that it can't call the function and listing all tried candidates.

Reply:
This error:
```
<source>:26:28: error: use of deleted function 'int foo::begin() const'

   26 |     std::as_const(f).begin();

      |                            ^

<source>:10:9: note: declared here

   10 |     int begin() const = delete;

      | 
```

vs this error:

```
<source>:27:28: error: passing 'std::add_const_t<bar>' {aka 'const bar'} as 'this' argument discards qualifiers [-fpermissive]

   27 |     std::as_const(b).begin();

      |                            ^

<source>:15:9: note:   in call to 'int bar::begin()'

   15 |     int begin()

      |         ^~~~~

Compiler returned: 1
```

From https://godbolt.org/z/a0oG8F